First sprint from AtolCD

ftavin opened the pull request:
Mantis issues:
- adding team members (#620)
- adding secondary OrgUnits (#243)
- adding framework in db (#106)

Tools for devs:
- adding docker-compose.yml file
- removal of unuseful permutations during maven compilation with sigmah-dev profile

ftavin committed:
Adding docker-compose.yml file for postgreSQL dev

aponcon-atolcd committed:
Add a mailhog docker image for managing emails

Change-Id: I0281a1aec3862691e70b98f07c4358775cb85307

ftavin committed:
Adding .docker to .gitignore

aponcon-atolcd committed:
Introduce permissions in team members view

(refs #620)

For the moment, the team member view page rely on UserPermission data for
non owners as an automatic Hibernate filter is filtering on these data.

Change-Id: I17a6e7768196308fb56412f4b6672ff9f1e9f8fe

aponcon-atolcd committed:
Add a tab for displaying project team members

(refs #620)

Change-Id: I5f06a6bcf21e4a1cdfaa2995ea87441907ee4ce6

aponcon-atolcd committed:
Add a button for removing a team member and a button for saving it

(refs #620)

Change-Id: Ic1c025b78e2c84866ba88f4361241e52a911951e

aponcon-atolcd committed:
Add default team member profiles form in project model administration

(refs #620)

Change-Id: I5713e98efccfacb6a2992e3959e130edf9d18a62

aponcon-atolcd committed:
Add a "Add team member" button to Project team member page

(refs #620)

Change-Id: If9268f6e8dfb82366f9b8fd11007a0dda6e49333

aponcon-atolcd committed:
Add a "Add team members by profile" button to Project team member page

(refs #620)

Change-Id: If695fe64c68fa0eea8d7d40dbc1e510dc966c539

aponcon-atolcd committed:
Fix a NPE when trying to update a user profile

Change-Id: Icdf69816553291e54b88643c96713448c88bc93a

aponcon-atolcd committed:
Separate VIEW_PROJECT and EDIT_PROJECT into *_ALL_PROJECTS and *_MY_PROJECTS

(refs #620)

Change-Id: Ic3aa1b7e802d385f22460007cc32963a2a5b7423

aponcon-atolcd committed:
Fix NPE when running the app with an old version of Sigmah

(refs #620)

Change-Id: I7ecc4e86f205cabbd8aea6eb2225619c804d7adc

aponcon-atolcd committed:
Disable an old activityinfo hibernate filter for team members queries

(refs #620)

Change-Id: I45af397fb697f677c36a775cff9d933a12eb5990

aponcon-atolcd committed:
Fix error when creating a new project with default team member profiles

(refs #620)

Change-Id: Ib1b9d0bbd729bf1f082306c1ac49c86a070d9de7

aponcon-atolcd committed:
Add a helper for getting the OrgUnit of a project

The problem is that a project can have several OrgUnits in the database.
But concretely, the application doesn't allow this situation and the code
is already making the assumption that a project has only one OrgUnit.
But the mean to get the project's OrgUnit was not very clean so let's
make it easier.

Change-Id: Ifc628926273268b038d9c4ea1887dd9c22a16e42

aponcon-atolcd committed:
Introduce the UI for adding secondary UserUnits

(refs #243)

Change-Id: I750df03e831a367881879bfac71f444183352a5e

aponcon-atolcd committed:
Fix user retrieval by org unit when in Offline Mode

(refs #620)

Change-Id: I2571f0f838ca05bad20ebf58cdd25d0ffaa99cd1

aponcon-atolcd committed:
Fix save action in TeamMember page when in Offline Mode

(refs #620)

Change-Id: I4396ed81ef69c88be4726f67872984b86bf0f761

aponcon-atolcd committed:
Remove save button in project team members page

(refs #620)

The data is saved each time a user or a profile is added to or removed from
the list.

Change-Id: I57481fb7b9f18726ac92628122066a2f2a0253e9

aponcon-atolcd committed:
Replace team member removal button by a clickable icon

(refs #620)

Change-Id: I1e48ffce340270850640fe53dd820623339718e6

aponcon-atolcd committed:
Verify framework fulfillment for project model related to a framework

(refs #106)

Change-Id: I8e536f91e393b524c25dc46e4d4f6639ed40765f

aponcon-atolcd committed:
Removal of unuseful permutations during maven compilation with sigmah-dev profile

By running 'mvn package -P sigmah-dev' instead of 'mvn package', the number
of permutations is hugely decreased : 2 permutations instead of 24.
For example, the compilation time is reduced to 1 minute instead of 4
minutes on my laptop.
The produced WAR has its size also reduced : 79MB instead of 124MB.
This improvement is useful for developers who want to produce a WAR and
upload it to a Tomcat but also if there is a Jenkins server which compiles
the project after each commit in Gerrit.

Change-Id: If469849ad2bd514ccd12216c0b694ebccaed7aad

aponcon-atolcd committed:
Fix offline mode for team member page

(refs #620)

Change-Id: Id72f2d058cefe17b7283e83d5f1013b0694b12d3

aponcon-atolcd committed:
Fix conflicts during load of projects in dashboard

Change-Id: Iacf3022b554aa8ddff3be699fa5384175c76e6f8

ftavin committed:
Fix retrieval of org units in admin view

Change-Id: I1093ab62accfd04b1783d3cbf841cc6bb99589e3

ftavin committed:
Add Offline feature for user units

(refs #243)

Change-Id: I75171e7e9dec5696b49455582794964d6599ddaa

ftavin committed:
Add a switch in the dashboard for hiding projects related to secondary org units

(refs #243)

Change-Id: I547846f07e5bb37ef4cc52ec1901c081d625dbc1

ftavin committed:
Add offline mode for "Add user profile to team" button in project team members page

(refs #620)

Change-Id: Ie36bf7ddd5e534b5c4e37f5b975f13a13917b0c5

ftavin committed:
Introduce secondary user units in the data model

(refs #243)

Change-Id: Ibc6ee1d7c1a5920bc7bce75600a601627494def8

ftavin committed:
Offline management for team members

(refs #620)

Change-Id: I1d38410f6c43fd5f8cc5883e2727c9c2012ab901

ftavin committed:
Fix of permission hierarchy and of permission goals

(refs #620)

Change-Id: I558a0f248ae15b6da2e0eb55d3b874b6d2259546

aponcon-atolcd committed:
Add Licence in file header

Change-Id: Iec8914642bd90166b61e32b9f8eea96c961347cd

aponcon-atolcd committed:
Fix secondary org units management

Change-Id: Iad9b9e643c95dad7cb927508e6d11220559b1d72

aponcon-atolcd committed:
Fix JSNI error with Profiler

Change-Id: I8b6024a8fead539f8e70f67186baa5da52affc31

aponcon-atolcd committed:
Fix MonitoredPoints and Reminders handlers

Change-Id: Ie4efb130cfdde769a4ec8946398202c03838b64f

aponcon-atolcd committed:
Fix of not complete user list in team member page in offline mode

Change-Id: I5fc2bdf67321e0a74eaeb971343dee1980795237

ftavin committed:
Fix visibility verification of projects

(refs #620)

Change-Id: Ieecf470ffe7098a60ae1ce25e2e9c275f654ec6b

aponcon-atolcd committed:
Fix of team member permissions in the project page

Change-Id: I5800185816c9ede1437244b113d4d52802434cb3

aponcon-atolcd committed:
Fix NPE while trying to update an orgUnit

Change-Id: Icaf02a9705f3b8eef52d1d05817a66fa9b6f715c

aponcon-atolcd committed:
Fix countries cache warming while running dev mode

Previously, the cache was loaded from data in indexedDB. As the dev mode
doesn't have access to indexedDB, it means the countries cache was never
loaded.

Change-Id: If3b1ee2876c0315ec5ceeb96a2a7c1989f44a430

aponcon-atolcd committed:
Prepare the ListComboBox to use it in contact details tab

Add the disable state
Allow to bring your own available value store.

(refs #634)

Change-Id: I98cfa5176fea33f0ed154424e82ed44ac40376b1

aponcon-atolcd committed:
Fix the tooltip in ListComboBox when there isn't any available value

Change-Id: Ie562ac3f7c1acb3fb234a515788c359ae76322d3

ftavin committed:
Introduce "framework" entities for ProjectModels

(refs #106)

Change-Id: I5e8b99679c929b1207208ad685962c2e34685ca3

numero-six commented on src/test/resources/META-INF/persistence.xml:
```diff
@@ -156,6 +156,8 @@
 		<class>org.sigmah.server.domain.profile.GlobalPermission</class>
 		<class>org.sigmah.server.domain.profile.OrgUnitProfile</class>
+		<class>org.sigmah.server.domain.profile.MainOrgUnitProfile</class>
+		<class>org.sigmah.server.domain.profile.SecondaryOrgUnitProfile</class>
```
It seems we have a problem here: those two classes does not exist. I guess these have been added at one point in the development process but removed later and someone forget to remove these two lines while removing the classes themselves.

@ftavin, as the author of the commit, can you confirm? (and commit a fix)

numero-six replied:
Also, it seems that 5 classes should be declared in this file while they are not (`org.sigmah.server.domain.Framework*`)

ftavin committed:
Fix : updating classes list in test/.../persistence.xml

Change-Id: I297adf4b38116c9c169e429f7eac920860fbe867

ftavin commented:
Class list fixed.

ghost commented:
Thanks!

There is a test failure which is specifically on the branch which is not on `master`:

```
-------------------------------------------------------------------------------
Test set: org.sigmah.server.servlet.importer.AutomatedImporterTest
-------------------------------------------------------------------------------
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.423 sec <<< FAILURE! - in org.sigmah.server.servlet.importer.AutomatedImporterTest
testImportCorrespondances(org.sigmah.server.servlet.importer.AutomatedImporterTest)  Time elapsed: 1.857 sec  <<< ERROR!
java.lang.NullPointerException: null
    at org.sigmah.server.handler.UpdateProjectHandler.searchForConflicts(UpdateProjectHandler.java:734)
    at org.sigmah.server.handler.UpdateProjectHandler.updateProject(UpdateProjectHandler.java:184)
    at com.google.inject.persist.jpa.JpaLocalTxnInterceptor.invoke(JpaLocalTxnInterceptor.java:66)
    at org.sigmah.server.handler.UpdateProjectHandler.execute(UpdateProjectHandler.java:148)
    at org.sigmah.server.handler.UpdateProjectHandler.execute(UpdateProjectHandler.java:95)
    at org.sigmah.server.handler.base.AbstractCommandHandler.execute(AbstractCommandHandler.java:117)
    at org.sigmah.server.dispatch.impl.UserDispatch.doExecute(UserDispatch.java:341)
    at org.sigmah.server.dispatch.impl.UserDispatch.access$000(UserDispatch.java:57)
    at org.sigmah.server.dispatch.impl.UserDispatch$UserExecutionContext.execute(UserDispatch.java:133)
    at org.sigmah.server.dispatch.impl.UserDispatch$UserExecutionContext.execute(UserDispatch.java:123)
    at org.sigmah.server.servlet.importer.AutomatedImporter.updateContainerWithDetails(AutomatedImporter.java:312)
    at org.sigmah.server.servlet.importer.AutomatedImporter.onContainerFound(AutomatedImporter.java:140)
    at org.sigmah.server.servlet.importer.AutomatedImporter.importCorrespondances(AutomatedImporter.java:102)
    at org.sigmah.server.servlet.importer.AutomatedImporterTest.testImportCorrespondances(AutomatedImporterTest.java:130)
```

The stack-trace leads to a code added in 11c3fc8980e3693c123f7be5cd8ef426a742bda2:

``` java
final ProfileDTO profile = Handlers.aggregateProfiles(context.getUser(), mapper).get(project.getOrgUnit().getId());
```

Didn't see anything obviously wrong at first sight, @ftavin can you check it? Thanks.

aponcon-atolcd committed:
Fix NPE during update project process

Make sure that a draft project (without org unit) can be updatable.

Change-Id: Id0e541dc410dc1b33e8f854f0778778d21e05b1d

ftavin commented:
Fixed !

aponcon-atolcd committed:
Initialize lists in Framework and TeamMember related entities

Change-Id: Ifed5fac865f562bcb314aceb264f8baa2af9b302

aponcon-atolcd committed:
Fix bad condition during verification of user rights for an OrgUnit

Change-Id: I63b037abc30bab9732eef0bce25256769cfe27b6

aponcon-atolcd committed:
Fix AutomatedImporterUnlockTest

Hibernate doesn't appreciate immutable lists. Let's replace immutable lists
(= Collections.singleton(...) or even Arrays.asList(...)) by good old
ArrayLists.
Also add a type to the newly created OrgUnitProfile.

Change-Id: I0d7f3d45259e603f83f7843b17c7f40dfe2e8729

aponcon-atolcd committed:
Fix AutomatedImportUnlockTest cleanup process

Change-Id: Ibd2ddcf4e717d6bf72dc21033200e25ad89997ed

ftavin commented:
New fix : we now have exactly the same test failure than the master branch !

ftavin commented:
Rebased on master in order to pass the checks

numero-six commented on src/main/java/org/sigmah/client/ui/presenter/DashboardPresenter.java:
```diff
@@ -22,464 +22,513 @@
+		});
+	}
+
+	Set<Integer> getOrgUnitIds() {
```
should be `private`

numero-six commented on src/main/java/org/sigmah/offline/dao/ProfileAsyncDAO.java:
```diff
@@ -0,0 +1,51 @@
+import org.sigmah.offline.js.ProfileJS;
+import org.sigmah.shared.dto.profile.ProfileDTO;
+
+public class ProfileAsyncDAO extends AbstractUserDatabaseAsyncDAO<ProfileDTO, ProfileJS> {
```
Indentation is 2 spaces, should be changed to tabs.

numero-six commented on src/main/java/org/sigmah/offline/dao/UserAsyncDAO.java:
```diff
@@ -182,6 +185,66 @@ public void onSuccess(OrgUnitDTO result) {
+		openTransaction(Transaction.Mode.READ_ONLY, new OpenTransactionHandler() {
+
+			@Override
+			public void onTransaction(final Transaction transaction) {
```
Generics are missing (compilation warning). I guess it should be
```java
new OpenTransactionHandler<Store>()
```
and 

```java
Transaction<Store> transaction
```

numero-six commented on src/main/java/org/sigmah/offline/dao/UserAsyncDAO.java:
```diff
@@ -182,6 +185,66 @@ public void onSuccess(OrgUnitDTO result) {
+	public void getByOrgUnits(final Set<Integer> orgUnitIds, final AsyncCallback<ListResult<UserDTO>> callback) {
+		final List<UserDTO> users = new ArrayList<UserDTO>();
+		// `i` needs to be final as it is accessed asynchronously
+		final int[] i = {0};
```
Will "i" is quite a clumsy variable name :wink: 

I guess you are trying to do a closure but since you want to update the integer but you can't since it's final, you circumvented that by using a array. If i'm right, could you replace that by using an instance of `org.apache.commons.lang3.mutable.MutableInt`.

```java
final MutableInt actualSize = new MutableInt();
```

and in the callback, replace

```java
i[0]++
```

by

```java
actualSize.increment();
```

Raphcal replied:
@numero-six I think that `org.apache.commons.lang3.mutable.MutableInt` is not accessible client-side.

numero-six replied:
@Raphcal do you have an idea of quick-fix that would be client-side compliant? Or maybe workaround? (well, at least, we could let that be a good variable name + a suitable comment)

Raphcal replied:
@numero-six Maybe we can introduce a simpler `MutableInt` of our own. I think it will be useful in other places of the project as well.

numero-six commented on src/main/java/org/sigmah/offline/handler/GetOrgUnitsAsyncHandler.java:
```diff
@@ -0,0 +1,89 @@
+		}
+
+		final List<OrgUnitDTO> orgUnitDTOs = new ArrayList<OrgUnitDTO>();
+		final int[] requests = new int[]{0};
```
Use MutableInt too

numero-six commented on src/main/java/org/sigmah/offline/handler/GetProfilesAsyncHandler.java:
```diff
@@ -0,0 +1,53 @@
+import org.sigmah.shared.command.result.ListResult;
+import org.sigmah.shared.dto.profile.ProfileDTO;
+
+public class GetProfilesAsyncHandler implements AsyncCommandHandler<GetProfiles, ListResult<ProfileDTO>>,
```
File should be indented with tabs

numero-six commented on src/main/java/org/sigmah/offline/inject/OfflineModule.java:
```diff
@@ -128,15 +71,20 @@ protected LocalDispatchServiceAsync provideLocalDispatcher(DispatchAsync dispatc
 			GetMonitoredPointsAsyncHandler getMonitoredPointsAsyncHandler,
 			GetOrganizationAsyncHandler getOrganizationAsyncHandler,
 			GetOrgUnitAsyncHandler getOrgUnitAsyncHandler,
+		GetOrgUnitsAsyncHandler getOrgUnitsAsyncHandler,
```
Indentation issue

numero-six commented on src/main/java/org/sigmah/offline/js/AuthenticationJS.java:
```diff
@@ -155,34 +169,114 @@ public native void setOrganizationLogo(String organizationLogo) /*-{
+		Set<Integer> orgUnitIds = new HashSet<Integer>();
+		JsArrayInteger secondaryOrgUnitIds = getSecondaryOrgUnitIds();
+		if (secondaryOrgUnitIds == null) {
+			return Collections.emptySet();
```
It's strange to declare and initialize `orgUnitIds` before returning `Collections.emptySet()`. Move `orgUnitIds` declaration after the if or return it line 184.

numero-six commented on src/main/java/org/sigmah/offline/js/CommandJS.java:
```diff
@@ -106,6 +101,9 @@ public static CommandJS toJavaScript(Command command) {
 			commandJS.setCommandType(Type.PREPARE_FILE_UPLOAD);
+		} else if(command instanceof UpdateProjectTeamMembers) {
+			commandJS = UpdateProjectTeamMembersJs.toJavascript((UpdateProjectTeamMembers) command);
+			commandJS.setCommandType(Type.UPDATE_PROJECT_TEAM_MEMBERS);
```
Shouldn't that structure be closed by a final `else` that throw an exception?

numero-six commented on src/main/java/org/sigmah/server/dao/impl/FrameworkFulfillmentHibernateDAO.java:
```diff
@@ -0,0 +1,46 @@
+			"JOIN pm.frameworkFulfillments ff " +
+			"JOIN FETCH ff.frameworkElementImplementations fei " +
+			"WHERE pm.id = :projectModelId " +
+			"AND ff.framework.availabilityStatus = 'AVAILABLE' ",
```
For compilation safety, `'AVAILABLE'` should not be hard-coded in the hql string but should be passed as a query argument

```
ff.framework.availabilityStatus = :available
```

then

```java
setParameter("available", Framework.AvailabilityStatus.AVAILABLE);
```

numero-six commented on src/main/java/org/sigmah/server/dao/impl/FrameworkHibernateDAO.java:
```diff
@@ -0,0 +1,62 @@
+			"SELECT f " +
+			"FROM Framework f " +
+			"JOIN FETCH f.frameworkHierarchies fh " +
+			"WHERE f.availabilityStatus = 'AVAILABLE' " +
```
Replace by HQL argument (see comment on FrameworkFulfillmentHibernateDAO)

numero-six commented on src/main/java/org/sigmah/server/dao/impl/FrameworkHibernateDAO.java:
```diff
@@ -0,0 +1,62 @@
+			"LEFT JOIN f.frameworkHierarchies fh " +
+			"LEFT JOIN fh.frameworkElements fe " +
+			"WHERE pm.id = :projectModelId " +
+			"AND f.availabilityStatus = 'AVAILABLE' " +
```
Replace by HQL argument (see comment on FrameworkFulfillmentHibernateDAO)

Raphcal replied:
@numero-six By a JPQL argument.

numero-six commented on src/main/java/org/sigmah/server/dao/impl/MonitoredPointHibernateDAO.java:
```diff
@@ -33,4 +38,23 @@
 
+	@Override
+	public List<MonitoredPoint> findNotCompletedByProjectIds(Set<Integer> projectIds) {
+		if (projectIds == null || projectIds.isEmpty()) {
```
Remove `projectIds == null || `, should never occur (I've checked the call and collection passed is never null, we should always expect empty collection is favored over null as a global rule of thumb).

numero-six commented on src/main/java/org/sigmah/server/dao/impl/ProjectHibernateDAO.java:
```diff
@@ -77,4 +79,70 @@
+	}
+
+	@Override
+	public List<Project> findProjectByTeamMemberIdAndOrgUnitIds(Integer userId, Set<Integer> orgUnitIds) {
```
could that be factored with the above method given that only the select clause is different?

numero-six commented on src/main/java/org/sigmah/server/dao/impl/ReminderHibernateDAO.java:
```diff
@@ -22,15 +22,38 @@
 public class ReminderHibernateDAO extends AbstractDAO<Reminder, Integer> implements ReminderDAO {
+	@Override
+	public List<Reminder> findNotCompletedByProjectIds(Set<Integer> projectIds) {
+		if (projectIds == null || projectIds.isEmpty()) {
```
Remove `projectIds == null || `

numero-six commented on src/main/java/org/sigmah/server/dao/impl/UserHibernateDAO.java:
```diff
@@ -86,39 +92,102 @@ public User findUserByChangePasswordKey(final String key) {
-		final StringBuilder query = new StringBuilder();
+	@Override
+	public List<User> findUsersByOrgUnitIds(Set<Integer> orgUnitIds) {
+		if (orgUnitIds == null || orgUnitIds.isEmpty()) {
```
Remove `orgUnitIds == null || `

numero-six commented on src/main/java/org/sigmah/server/dao/impl/UserUnitDAOImpl.java:
```diff
@@ -46,26 +49,75 @@
+				"JOIN FETCH oup.profiles p " +
+				"JOIN FETCH oup.orgUnit o " +
+				"WHERE oup.user.id = :userId " +
+				"AND oup.type = 'MAIN' ",
```
Use typed query argument instead of hard-coded enum name

numero-six commented on src/main/java/org/sigmah/server/domain/Project.java:
```diff
@@ -22,338 +22,373 @@
+		for (OrgUnit orgUnit : getPartners()) {
+			return orgUnit;
+		}
+		return null;
```
Should we ever get there? Shouldn't we throw an exception if partners is empty?

numero-six commented on src/main/java/org/sigmah/server/handler/GetMonitoredPointsHandler.java:
```diff
@@ -102,52 +117,45 @@
+		for (MonitoredPoint monitoredPoint : monitoredPoints) {
+			MonitoredPointDTO monitoredPointDTO = mapper().map(monitoredPoint, new MonitoredPointDTO(), mappingMode);
+
+			// FIXME: Put this code into a DAO
```
FIXME should be fixed (or removed if obsolete).

numero-six commented on src/main/java/org/sigmah/server/handler/GetProjectTeamMembersHandler.java:
```diff
@@ -0,0 +1,82 @@
+		}
+
+		List<Profile> profiles = profileDAO.getProjectTeamMemberProfiles(command.getProjectId());
+		List<ProfileDTO> profileDTOs = new ArrayList<>();
```
Just a comment: i've seen several places where a List is dynamically filed (using List#add in a loop) and dynamically read (via foreach) without any random access read (ie `List#get(index)`). In this case, `LinkedList` should be favored over `ArrayList`.

numero-six commented on src/main/java/org/sigmah/server/handler/UpdateProjectHandler.java:
```diff
@@ -163,21 +164,32 @@ public VoidResult execute(final UpdateProject cmd, final UserExecutionContext co
-		
+		if (project != null) {
+			if (!Handlers.isProjectEditable(project, user)) {
+				throw new IllegalStateException();
```
Could you please add a message everytime you throw this exception (and a different one for each throw clause)? It would ease analyzing the stack-trace even if the code have moved (4 `throw` clause to fix).

I will update the coding guidelines for that.

numero-six commented on src/main/java/org/sigmah/server/handler/UpdateProjectTeamMembersHandler.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.List;
+import java.util.Set;
+
+import javax.inject.Inject;
```
Well, technically this is OK but I would favor consistency all over the code base, could you replace (multiple occurrences in the PR)

```java
import javax.inject.Inject;
```

with

```java
import com.google.inject.Inject;
```

which have 285 occurrences so far. Or maybe we could migrate everything to `javax.inject`, @Raphcal any feedback about that?

numero-six commented on src/main/java/org/sigmah/server/handler/UpdateProjectTeamMembersHandler.java:
```diff
@@ -0,0 +1,102 @@
+
+		List<ProfileDTO> profileDTOs = new ArrayList<>();
+		for (Profile profile : project.getTeamMemberProfiles()) {
+			ProfileDTO profileDTO = mapper().map(profile, ProfileDTO.class, ProfileDTO.Mode.BASE);
```
Call to `map` is deprecated for performance reasons. Is it easy to replace those calls by non-deprecated API calls. Since some people are working on the performance issues... we may not add complications.

numero-six commented on src/main/java/org/sigmah/server/handler/UpdateProjectTeamMembersHandler.java:
```diff
@@ -0,0 +1,102 @@
+		Set<Integer> teamMemberIds = new HashSet<>();
+		Set<Integer> teamMemberProfileIds = new HashSet<>();
+
+		// TODO: Verify that the users are related to the org unit of the project
```
those 2 `TODO`s should be done or removed.

numero-six commented on src/main/java/org/sigmah/server/handler/util/Handlers.java:
```diff
@@ -22,264 +22,355 @@
+					}
+				}
+			}
+		}
```
Indentation is broken and does not reflect imbrication levels.

numero-six commented on src/main/java/org/sigmah/server/handler/util/Handlers.java:
```diff
@@ -22,264 +22,355 @@
+		return isProjectAccessible(project, user, true);
+	}
+
+	public static boolean isProjectAccessible(Project project, User user, boolean edition) {
```
Great! I like providing two method instead of a flag argument for clarity. So, this method should be private to force call on one of the two above methods.

numero-six commented on src/main/java/org/sigmah/server/handler/util/Handlers.java:
```diff
@@ -22,264 +22,355 @@
+					targetedOrgUnitProfile = orgUnitProfile;
+					minDistance = currentDistance;
+					break;
+		}
```
Indentation broken, does not reflect the imbrication levels.

numero-six commented on src/main/java/org/sigmah/server/security/impl/AccessRights.java:
```diff
@@ -22,413 +22,363 @@
+			if (grantType != null && grantType == GrantType.ANONYMOUS_ONLY) {
+				return false;
+			} else {
+				for (Map.Entry<Integer, ProfileDTO> aggregatedProfileEntry : Handlers.aggregateProfiles(user, mapper).entrySet()) {
```
Instead of `entrySet()`, `values()` should be called since the key of the entry is not needed in the body of the loop.

numero-six commented on src/main/java/org/sigmah/server/service/ProjectModelService.java:
```diff
@@ -22,484 +22,503 @@
+				model.setDateMaintenance(null);
+			}
+		}
+		if (changes.containsKey(AdminUtil.PROP_PM_DEFAULT_TEAM_MEMBER_PROFILES)) {
```
Three things:

- `PropertyMap#get` is a generic method that deal with the cast for you. So should declare et `Set<Integer>` instead of an `Object`
- `findByIds` returns an empty collection for an empty id collection
- ` ! defaultProfileIds instanceof Set` should never occur (given the code, set is always called with a Set<Integer>`)

So, it seems that this part (line 277 to 289) could be replaced by simply

```java
if (changes.containsKey(AdminUtil.PROP_PM_DEFAULT_TEAM_MEMBER_PROFILES)) {
	Set<Integer> defaultProfileIds = changes.get(AdminUtil.PROP_PM_DEFAULT_TEAM_MEMBER_PROFILES);
	List<Profile> profiles = profileDAO.findByIds(defaultProfileIds);
	model.setDefaultTeamMemberProfiles(profiles);
}
```

numero-six commented on src/main/java/org/sigmah/server/service/UserPermissionPolicy.java:
```diff
@@ -99,30 +100,40 @@ public UserPermissionPolicy(UserPermissionDAO userPermissionDAO, UserDAO userDAO
 	 * Each time a user is created or modified a method will go through all projects which are editable by the user
 	 * (projects attached to the user main orgunit, and all the child orgunits of it), and add a row in the table
-	 * userpermission for each of them if the user has the EDIT_PROJECT global privilege
+	 * userpermission for each of them if the user has the EDIT_ALL_PROJECTS global privilege
```
When referencing a constant, please use the appropriate Javadoc markup so when refactoring (deleting or renaming the constant), the IDE may alert you to update the Javadoc or update it for you.

Replace by ```{@link GlobalPermissionEnum#EDIT_ALL_PROJECTS}``` (other occurrences in the PR).

numero-six commented on src/main/java/org/sigmah/server/service/UserService.java:
```diff
@@ -1,280 +1,314 @@
+		if (userToPersist.getId() != null) {
+			// Needs to remove all subs profiles to avoid orphan entities
+			while (!userToPersist.getOrgUnitsWithProfiles().isEmpty()) {
+				userToPersist.getOrgUnitsWithProfiles().get(0).getProfiles().removeAll(userToPersist.getOrgUnitsWithProfiles().get(0).getProfiles());
```
Is there a reason why this is not simply

```java
userToPersist.getOrgUnitsWithProfiles().get(0).getProfiles().clear();
```
?

numero-six commented on src/main/java/org/sigmah/server/service/UserService.java:
```diff
@@ -1,280 +1,314 @@
+
+				OrgUnitProfile secondaryOrgUnitProfile = createOrUpdateUserUnit(userUnitDTO, userToPersist);
+				userToPersist.getOrgUnitsWithProfiles().add(secondaryOrgUnitProfile);
+				}
```
Identation

numero-six commented on src/main/java/org/sigmah/shared/command/result/Authentication.java:
```diff
@@ -177,16 +181,16 @@ public Authentication(Integer userId, String userEmail, String userName, String
-	 * @param orgUnitId
+	 * @param mainOrgUnitId
 	 *          The user's OrgUnit id.
 	 * @param aggregatedProfile
```
Unknown param, I guess it's a missing "s" ;)

numero-six commented on src/main/java/org/sigmah/shared/command/result/Authentication.java:
```diff
@@ -217,8 +222,31 @@ public String toString() {
+			}
+		}
+		serializedAggregatedProfiles.append("]");
+		builder.append("aggregatedProfiles", serializedAggregatedProfiles);
```
This code is quite understandable but if this code of construction occur elsewhere, please introduce a utility class or just use something like Guava's com.google.common.base.Joiner

```java
Joiner.on(", ").withKeyValueSeparator(": ").join(aggregatedProfiles);
```

numero-six commented on src/main/java/org/sigmah/server/service/UserPermissionPolicy.java:
```diff
@@ -228,4 +238,24 @@ public boolean isGranted(final OrgUnitProfile userOrgUnit, final GlobalPermissio
 		return false;
 	}
 
+	public boolean isGranted(List<OrgUnitProfile> userUnits, GlobalPermissionEnum permission) {
```
I guess parameters should have been declared `final` to be consistent with above methods.

@Raphcal I know that you are a strong advocate of `final` everything (i like it), but when browsing the code base is not consistent. I was wondering, do you have a rule of when to declare things `final` or is it just that some guys did not followed the practice?

Raphcal replied:
@numero-six It is usually the latter. I love `final` but we didn't really enforce its usage so a lot of methods are missing it.

ghost commented:
Woaw, I've made tens of comments :open_mouth: 

@ftavin no panic! I'm globally quite happy with the code and I think most of the things I've noted are secondary details that should not prevent a merge. Most of them should be fixed in seconds.

I let @osarrat see, according to the time AtolCD have and priorities, if the lil' time to spend worth it.

osarrat commented:
If @numero-six , you consider those changes secondary, I let @ftavin free to decide how much it is worth to work on them now. If @ftavin prefers to postpone or not work on it because workplan is too tight until release, then I will open an issue on next-but-one version to work on this.

ghost commented:
Merging to meet deadlines but fixes of above comments will still be welcomed
